### ISSUE TITLE
MiniMax Music 3 modular pipeline: 2.9x consumer-GPU speedup - findings that may be worth upstreaming

### ISSUE THREAD

**Issue by TheDutchRuler** [huggingface/diffusers#14486]
Following the merged MiniMax Music 3 modular pipeline (#14456) — thank you for
it — we rebuilt its hot paths for single-GPU speed and open-sourced the result:

https://github.com/TheDutchRuler/minimax-music3-studio

Measured on an RTX 4090 (20s songs, warm, bf16): reference 50.5s/song →
17.7s/song (0.88x realtime) with quality-neutral changes only. Findings that
may be worth upstreaming or documenting:

1. **`torch.compile` has a silent no-op trap.** The semantic-generation block
   calls `language_model.model(...)` (the submodule) in its decode loop, so
   `torch.compile(pipe.language_model)` compiles a callable the loop never
   invokes. Compiling the wrong target costs nothing and gains nothing —
   easy to burn time on.

2. **A manually built `StaticCache` never gets CUDA graphs.** transformers'
   compiled-generate path calls `torch._dynamo.mark_static_address` on the
   cache tensors; building the cache by hand skips that, and cudagraph trees
   silently refuse to capture the in-graph cache mutation (no warning at
   default log levels). Marking the tensors after prefill fixes it.

3. **Cheap wins in the blocks themselves:**
   - The DiT inner loop runs conditional/unconditional as two sequential
     batch=1 forwards; one batch=2 forward is formula-identical (12.8s → 8.8s
     for the stage in our measurements).
   - `lm_head` reads a 1.38GB matrix per frame while every non-audio row is
     masked to -inf; slicing to the sampleable rows is provably identical.
   - Replacing the final softmax+multinomial with Gumbel-argmax (identical
     distribution, unit-tested at L1≈0.015 over 200k draws) lets the whole
     depth-decoder chain compile with zero eager glue.

4. **The big one — batched variation generation.** The AR stage is memory-
   bandwidth bound (~23GB weight reads/frame), so K same-prompt variations
   decoded in lockstep (batch 2K rows, per-song seeded noise, cross-song
   batched CFG denoise grouped by exact frame count) make the marginal song
   nearly free: 3 variations at 17.7s each vs 31s solo. Everything stays
   row-independent; might be worth first-class support in the pipeline.

Also documented: negative results (per-layer group offload on the AR LLM
re-streams 16.4GB/frame — the model card's low-VRAM snippet; FP8 weight-only
via torchao 2.1x slower on Windows/torch 2.11), details and harnesses in the
repo's `labs/`.

This work was engineered end-to-end with Claude (Fable 5 Max) by Anthropic.
Happy to expand any of these into a proper issue/PR if useful.

**Comment by yiyixuxu** [comment]
thanks @TheDutchRuler 
cc @asomoza @apolinario here

Can you verify the batched CFG performance claim? we have not seen that on other models, but if it's true for Minimax Music, I think we should include a new ChunkDenoiseInner block that does not use the guider, 

https://github.com/huggingface/diffusers/blob/main/src/diffusers/modular_pipelines/minimax_music3/denoise.py#L160

Maybe use this one in the default blocksets and document how to swap out to use the guider-equipped one for different guidance techniques

**Comment by yiyixuxu** [comment]
@TheDutchRuler 
if you can send a PR on this it'd be great!

>lm_head reads a 1.38GB matrix per frame while every non-audio row is
masked to -inf; slicing to the sampleable rows is provably identical.

**Comment by yiyixuxu** [comment]
@TheDutchRuler 
on this
>The big one — batched variation generation. The AR stage is memory-
bandwidth bound (~23GB weight reads/frame), so K same-prompt variations
decoded in lockstep (batch 2K rows, per-song seeded noise, cross-song
batched CFG denoise grouped by exact frame count) make the marginal song
nearly free: 3 variations at 17.7s each vs 31s solo. Everything stays
row-independent; might be worth first-class support in the pipeline.

can you host this as a hub custom block first? we can help to promote and may decide to upstream later 

cc @asomoza too

**Comment by TheDutchRuler** [comment]
@yiyixuxu Verification done, as requested — and your skepticism was well placed.

**Method:** one process, one model load, alternating ONLY the denoise inner
block between the reference implementation (guider, two sequential batch-1
transformer passes) and a batched one (single batch-2 pass, same guidance
formula), same seed each render, A→B→A→B ordering to expose warm-up drift,
DiT compilation disabled in both arms so batching is the only variable.
Stage times from block-level timers with CUDA synchronization; RTX 4090,
bf16, Windows. Harness: `labs/verify_batched_cfg.py` in the repo.

**Result (denoise stage, best-of-2 per arm, 20s song = 4 chunks x 30 steps):**

| | reference (guider) | batched CFG | speedup |
|---|---|---|---|
| denoise stage | 9.41s | 8.81s | **1.07x** |

Duration should not change this ratio: the DiT always operates on fixed
200-frame chunk windows, so per-step shapes are duration-invariant (a 60s
reference arm measured 29.36s ≈ 3.12x the 20s stage — pure chunk-count
scaling, consistent with that).

GPU utilization is 85-94% in BOTH arms — the batch-1 DiT forward is not as
underutilized as we assumed. The larger delta in our original issue numbers
(12.8s → 8.8s) came from measurements taken in different sessions during an
optimization ladder; the paired test shows most of that was warm-up settling,
not batching. Our mistake — the paired design you implicitly asked for was
the right check.

**Recommendation:** ~7% is probably not worth a guider-free default
`ChunkDenoiseInner` block and the API asymmetry it brings. If you still want
it as an opt-in block we're happy to contribute it, but we would not claim
more than ~5-10% for it.

**What does survive paired verification:**

- The lm_head slice: measured **5.31ms → 0.54ms per frame** (9.9x) on the
  head op — at 25 frames/s that is ~2.4s per 20s song. One precision note
  before the PR: slicing changes the GEMM tile shape, so logits on the
  sampleable rows differ from the full head by up to one bf16 ulp (0.03125
  measured) — mathematically the same dot products, identical distribution,
  but a given seed can take a different (equally valid) trajectory, as with
  any kernel-shape change. The PR will state this exactly. PR incoming.

- The ensemble (batched variations) numbers, which rest on a different
  mechanism: the autoregressive stage is weight-read bound at seq=1, so K
  variations share the ~23GB/frame read. Those were same-session sequential
  measurements (3x solo 31s vs one 53s trio). We'll package it as a Hub
  custom block per your suggestion.

Verification and this write-up were done with Claude (Fable 5 Max).

**Comment by TheDutchRuler** [comment]
Addendum — we dug into *why* batching tops out, to make sure we were not leaving anything on the table:

**DiT batch-scaling microbenchmark** (transformer only, real weights, seq=689, RTX 4090 bf16):

| batch | ms/sample | effective TFLOPS |
|---|---|---|
| 1 | 32.3 | 104 |
| 2 | 29.0 | 116 |
| 4 | 28.5 | 117 |
| 8 | 27.9 | 120 |

The DiT is compute-saturated from batch 1, so ~1.07-1.12x is the ceiling for ANY batching scheme on this stage (CFG or multi-sample) — consistent with your experience on other models. We would recommend against the guider-free default block.

The remaining kernel headroom is real but modest: `torch.compile(mode="max-autotune-no-cudagraphs")` measured **+7-8%** (117→127 effective TFLOPS) at the production shapes, after a one-off kernel search. We ship it as an opt-in.

(With Claude, Fable 5 Max.)

**Comment by TheDutchRuler** [comment]
@yiyixuxu Done — the ensemble is live as a Hub custom block:

**https://huggingface.co/Rreitsma/minimax-music3-ensemble-blocks**

```python
from diffusers.modular_pipelines import ModularPipelineBlocks, ComponentsManager
import torch

blocks = ModularPipelineBlocks.from_pretrained(
    "Rreitsma/minimax-music3-ensemble-blocks", trust_remote_code=True
)
manager = ComponentsManager()
manager.enable_auto_cpu_offload(device="cuda")
pipe = blocks.init_pipeline("MiniMaxAI/MiniMax-Music3", components_manager=manager)
pipe.load_components(dtype=torch.bfloat16)
audios = pipe(prompt=..., lyrics=..., audio_duration=60.0, num_variations=3, seed=7, output="audios")
```

Three blocks (reusing your `TokenizeStep`): a lockstep batched autoregressive step (batch 2K rows, per-variation seeded generators, reference sampling recipe, early-EOS rows freeze at no cost) and a grouped batched denoise+vocode step (variations with equal frame counts share the pass — no padding). Deliberately plain eager PyTorch, no extra dependencies, so it runs wherever diffusers main runs.

Measured with the published block (diffusers main, RTX 4090 bf16, warm, 15s songs):

| variations | total | per variation |
|---|---|---|
| 1 | 28.9s | 28.9s |
| 2 | 34.9s | 17.5s (1.65x) |
| 3 | 41.5s | **13.8s (2.1x)** |

The marginal variation costs ~21% of a solo render — consistent with the AR stage being weight-read bound. Caveats documented on the card (standard CFG only; VRAM scales with duration x variations; batched kernels differ from solo at ulp level so per-seed trajectories are not solo-reproducing).

Happy to iterate on naming/IO conventions if you want it closer to upstream shape. (Built with Claude, Fable 5 Max.)

**Comment by bghira** [comment]
@TheDutchRuler you could use torch profiler to see gaps and kernel launch count, which can be quite large even for regional compile. it's possible to make the entire denoiser block a single compiled function and that'll reduce more than 2400 kernel launches down to something like number of steps + a few instead.
